test/stack_controller_test.go

```diff
@@ -319,6 +340,59 @@ var _ = Describe("Stack Controller", func() {
 			return k8serrors.IsNotFound(err)
 		}, timeout, interval).Should(BeTrue())
 	})
+
+	It("Should deploy an AWS S3 Stack successfully with credentials passed through EnvRefs", func() {
+		var stack *pulumiv1alpha1.Stack
+		stackName := fmt.Sprintf("%s/s3-op-project/dev-env-ref-%s", stackOrg, randString())
+		fmt.Fprintf(GinkgoWriter, "Stack.Name: %s\n", stackName)
+
+		// Write a secret to a temp directory. This is a stand-in for other mechanisms to reify the secrets
+		// on the file system. This is not a recommended way to store/pass secrets.
+		Eventually(func() bool {
+			if err = os.WriteFile(filepath.Join(secretsDir, pulumiAPISecretName), []byte(pulumiAccessToken), 0600); err != nil {
+				return false
+			}
+			return true
+		}, timeout, interval).Should(BeTrue())
+
+		// Define the stack spec
+		spec := pulumiv1alpha1.StackSpec{
+			// Cover all variations of resource refs
+			EnvRefs: map[string]pulumiv1alpha1.ResourceRef{
+				"PULUMI_ACCESS_TOKEN":   pulumiv1alpha1.NewFileSystemResourceRef(filepath.Join(secretsDir, pulumiAPISecretName)),
+				"AWS_ACCESS_KEY_ID":     pulumiv1alpha1.NewLiteralResourceRef(awsAccessKeyID),
+				"AWS_SECRET_ACCESS_KEY": pulumiv1alpha1.NewSecretResourceRef(namespace, pulumiAWSSecret.Name, "AWS_SECRET_ACCESS_KEY"),
+				"AWS_SESSION_TOKEN":     pulumiv1alpha1.NewEnvResourceRef("AWS_SESSION_TOKEN"),
+			},
+			Config: map[string]string{
+				"aws:region": "us-east-2",
+			},
+			Stack:             stackName,
+			ProjectRepo:       "https://github.com/metral/test-s3-op-project", // TODO: relocate to some other repo
+			Commit:            "bd1edfac28577d62068b7ace0586df595bda33be",
+			DestroyOnFinalize: true,
+		}
+
+		// Create stack
+		name := "stack-test-aws-s3-file-secrets"
+		stack = generateStack(name, namespace, spec)
+		Expect(k8sClient.Create(ctx, stack)).Should(Succeed())
+
+		// Check that the stack updated successfully
+		Eventually(func() bool {
+			fetched := &pulumiv1alpha1.Stack{}
+			err := k8sClient.Get(ctx, types.NamespacedName{Name: stack.Name, Namespace: namespace}, fetched)
+			if err != nil {
+				return false
+			}
+			if fetched.Status.LastUpdate != nil {
+				return fetched.Status.LastUpdate.LastSuccessfulCommit == stack.Spec.Commit &&
+					fetched.Status.LastUpdate.LastAttemptedCommit == stack.Spec.Commit &&
+					fetched.Status.LastUpdate.State == pulumiv1alpha1.SucceededStackStateMessage
```

Comment:
If this is done commonly, it would be good to expose as a function. Fine for this test, if not.

Reply:
Done.